File: solutions/Python/2300.SuccessfulPairsOfSpellsAndPotions.py
from typing import List


class Solution:
    def successfulPairs(
        self, spells: List[int], potions: List[int], success: int
    ) -> List[int]:
        potions.sort()
        n = len(potions)
        result = []

        for spell in spells:
            required_potion = (success + spell - 1) // spell

            left = 0
            right = n
            while left < right:
                mid = (left + right) // 2
                if potions[mid] >= required_potion:
                    right = mid
                else:
                    left = mid + 1

            count = n - left
            result.append(count)

        return result

File: solutions/JavaScript/2300.SuccessfulPairsOfSpellsAndPotions.js
var successfulPairs = function (spells, potions, success) {
    potions.sort((a, b) => a - b);
    const n = potions.length;
    const result = [];

    for (const spell of spells) {
        const requiredPotion = Math.floor((success + spell - 1) / spell);

        let left = 0;
        let right = n;
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (potions[mid] >= requiredPotion) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        const count = n - left;
        result.push(count);
    }

    return result;
};